src/main/scala/com/gu/fezziwig/CirceScroogeWhiteboxMacros.scala

```diff
@@ -0,0 +1,301 @@
+      }
+    )
+    val reprType: Tree = params.foldRight[Tree](tq"""_root_.shapeless.HNil""") { case (param, acc) =>
+      val witnessName = TermName(s"${param.name}Witness")
+      val fieldType: Tree = if (param.asTerm.isParamWithDefault) {
+        tq"Option[${param.typeSignature}]"
+      } else {
+        q"${param.typeSignature}"
+      }
+      tq"""_root_.shapeless.::[_root_.shapeless.labelled.FieldType[${witnessName}.T, ${fieldType}], $acc]"""
+    }
+    val hlist = params.foldRight[Tree](q"""_root_.shapeless.HNil""") { case (param, acc) =>
+      val paramName = TermName(param.name.toString())
+      q"""_root_.shapeless.::(${paramName}, $acc)"""
+    }
+    // hlistPattern is identical to hlist but with the pq interpolator instead of q
+    val hlistPattern = params.foldRight[Tree](pq"""_root_.shapeless.HNil""") { case (param, acc) =>
+      val paramName = TermName(param.name.toString())
+      pq"""_root_.shapeless.::(${paramName}, $acc)"""
+    }
+    val labelledFields = params.map(param => {
+      val paramName = TermName(param.name.toString())
+      val witnessName = TermName(s"${param.name.toString()}Witness")
+      val (fieldExpr, fieldType): (Tree, Tree) = if (param.asTerm.isParamWithDefault) {
+        (q"_root_.scala.Some(struct.${paramName})", tq"Option[${param.typeSignature}]")
+      } else {
+        (q"struct.${paramName}", q"${param.typeSignature}")
+      }
+      q"""val ${paramName}: _root_.shapeless.labelled.FieldType[${witnessName}.T, ${fieldType}] = _root_.shapeless.labelled.field(${fieldExpr})"""
+      }
+    )
+    val labelledGeneric = q"""
+    new _root_.shapeless.LabelledGeneric[$A] {
+      type Repr = ${reprType}
+
+      def to(struct: $A): Repr = {
+        ..$labelledFields
+        $hlist
+      }
+
+      def from(hlist: Repr): $A = hlist match {
+        case $hlistPattern => $apply(..${fieldNames})
+      }
+    }"""
+    q"""{
+      ..$witnesses
+
+      $labelledGeneric
+    }"""
+  }
+
+  /**
+    * Provides a LabelledGeneric implicit for Thrift unions, as
+    * [[thriftStructGeneric]] does for structs.
+    *
+    * Given this thrift definition
+    *
+    * {{{
+    * union Union1 {
+    *   1: StructC c
+    *   2: StructD d
+    * }
+    * }}}
+    *
+    * the produced implicit will be roughly equivalent to
+    *
+    * {{{
+    * {
+    *   import shapeless.{Witness, :+:, labelled, CNil, Inl, Inr}
+    *
+    *   val dWitness = Witness(Symbol("d"))
+    *   val cWitness = Witness(Symbol("c"))
+    *   type Union1Repr = labelled.FieldType[dWitness.T, StructD] :+: labelled.FieldType[cWitness.T, StructC] :+: CNil
+    *
+    *   new LabelledGeneric[Union1] {
+    *     type Repr = Union1Repr
+    *
+    *     def to(union: Union1): Repr = union match {
+    *       case Union1.D(x) => Inl(labelled.field(x))
+    *       case Union1.C(x) => Inr(Inl(labelled.field(x)))
+    *       case Union1.UnknownUnionField(_) => throw new RuntimeException("Unknown union field")
```

Comment:
Should this handle unknown union fields better? How is this handled in the current implementation? I will check.

Reply:
Given the way we serialise I’m not certain how we would detect unknown union fields: I suppose we’d look for any keys that exist on the object and read the first one we find as the unknown union field?

Reply:
In the current implementation, the encoder serialises `UnknownUnionField` values as `null`. The decoder doesn’t ever produce a value of type `UnknownUnionField`, and returns a `DecodingFailure` for `null`.

Reply:
I also found a bug in the decoder macro, I think: it calls `Either.left` in it, which doesn’t exist. I’m pretty sure it’s meant to be `_root_.scala.util.Left` instead. I don’t understand why that only fails to typecheck if I explicitly materialise the decoder though – our tests successfully decode values of union types in fields of structs, which is a bit mysterious.

Reply:
I don’t think I can easily match the current behaviour: I’m not sure how to create a generic representation whose derived decoder will fail to decode what the derived encoder writes. But I don’t think having a non-roundtripping encoder/decoder pair is a good idea anyway.

So I think I should consider incorporating unknown fields into the representation. I guess this would be an extra `:+:` branch, either using a field name that’s likely to be uncommon or automatically picking a field name that’s unused by the union. I think the former is probably fine, as long as I pick a very awkward name – thinking about how to deterministically choose an unused name (so that it’s the same every time the generic is derived) makes me uneasy.

Reply:
I have now updated this PR to handle unknown fields by preserving the field names, types, and ids, but dropping the field values.